graphql_compiler/schema_transformation/split_query.py

```diff
@@ -0,0 +1,118 @@
+)
+
+from ..schema import OutputDirective
+
+
+QueryConnection = namedtuple(
+    'QueryConnection', (
+        'sink_query_node',  # SubQueryNode
+        'source_field_out_name',
+        # str, the unique out name on the @output of the the source property field in the stitch
+        'sink_field_out_name',
+        # str, the unique out name on the @output of the the sink property field in the stitch
+    )
+)
+
+
+class SubQueryNode(object):
+    def __init__(self, query_ast):
+        """Represents one piece of a larger query, targeting one schema.
+
+        Args:
+            query_ast: Document, representing one piece of a query
+        """
+        self.query_ast = query_ast
+        self.schema_id = None  # str, identifying the schema that this query targets
+        self.parent_query_connection = None
+        # SubQueryNode or None, the query that the current query depends on
+        self.child_query_connections = []
+        # List[SubQueryNode], the queries that depend on the current query
+
+
+def _get_output_directive(out_name):
+    """Return a Directive representing an @output with the input out_name."""
+    return Directive(
+        name=Name(value=OutputDirective.name),
+        arguments=[
+            Argument(
+                name=Name(value=u'out_name'),
+                value=StringValue(value=out_name),
+            ),
+        ],
+    )
+
+
+def _get_query_document(root_vertex_field_name, root_selections):
+    """Return a Document representing a query with the specified name and selections."""
+    return Document(
+        definitions=[
+            OperationDefinition(
+                operation='query',
+                selection_set=SelectionSet(
+                    selections=[
+                        Field(
+                            name=Name(value=root_vertex_field_name),
+                            selection_set=SelectionSet(
+                                selections=root_selections,
+                            ),
+                            directives=[],
+                        )
+                    ]
+                )
+            )
+        ]
+    )
+
+
+def _add_query_connections(parent_query_node, child_query_node, parent_field_out_name,
+                           child_field_out_name):
+    """Modify parent and child SubQueryNodes by adding QueryConnections between them."""
+    if child_query_node.parent_query_connection is not None:
+        raise AssertionError(
+            u'The input child query node already has a parent connection, {}'.format(
+                child_query_node.parent_query_connection
+            )
+        )
+    if any(
+        query_connection_from_parent.sink_query_node is child_query_node
+        for query_connection_from_parent in parent_query_node.child_query_connections
+    ):
+        raise AssertionError(
+            u'The input parent query node already has the child query node in a child query '
+            u'connection.'
```

Comment:
Would be nice to output both the parent and child nodes in both of these assertion error messages, it would greatly help debugging.

Reply:
Just outputting these nodes isn't super helpful since it'll just print SubQueryNode object at some location. I could include their ASTs in the error message?

Reply:
Right. However, decorating the class with `@six.python_2_unicode_compatible` and then defining `__str__` / `__repr__` methods would be a reasonable fix. The `Location` class in `helpers.py` is a good example of how this would work.

I think the ASTs would be good to have too, but I'd also like the other information, like `schema_id` etc.